Update to dotnet 6.0 LTS and latest version of battlesnake api

lvernaillen opened the pull request:
- Updated to next LTS version (.NET 6) because .NET Core 3.1 reaches end of support on December 13, 2022
  https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core
- Updated to latest BattleSnake API.
  See https://docs.battlesnake.com/api

lvernaillen committed:
Upgrade to net6.0

lvernaillen committed:
Upgrade to latest battlesnake api model

- Add documentation to request and response objects
- Renamed Point to Coordinate so same terminology is used in documentation and code
- Removed unnecessary tabs/spaces
- Add Source and Map to Game object
- Add RulesetSettings
- Add SnakeCustomizations to Snake object
- Snake Health is now int instead of string
- Snake Latency is now string instead of int

neistow commented on Starter.Api/Responses/InitResponse.cs:
```diff
@@ -2,10 +2,34 @@
+        /// </summary>
         public string ApiVersion { get; set; }
+
+        /// <summary>
+        /// Username of the author of this Battlesnake. If provided, this will be used to verify ownership.Example: "BattlesnakeOfficial"
+        /// </summary>
         public string Author { get; set; }
+
+        /// <summary>
+        /// Hex color code used to display this Battlesnake. Must start with "#" and be 7 characters long.Example: "#888888"
+        /// </summary>
         public string Color { get; set; }
+
+        /// <summary>
+        /// Displayed head of this Battlesnake. See Customization Guide for available optionsExample: "default"
+        /// </summary>
         public string Head { get; set; }
+
+        /// <summary>
+        /// Displayed tail of this Battlesnake. See Customization Guide for available optionsExample: "default"
+        /// </summary>
         public string Tail { get; set; }
+
+        /// <summary>
+        /// A version number or tag for your snake.
+        /// </summary>
+        public string Version { get; set; }
```
I think it would be good to add nullable annotations to emphasize the nullability of those fields according to the [docs](https://docs.battlesnake.com/api/requests/info)

lvernaillen replied:
I made them nullable as per the docs.
Also enabled nullable context, since that is the default in .NET 6.

neistow commented:
Also, since the project is moving to .NET 6 it would be good to bring all of the features including minimal API, to reduce project complexity and make it more friendly for beginners. This is not a change request but more of an enhancement proposal that may be done in the future by someone else.

lvernaillen committed:
Enable nullable context

- Beginning with .NET 6, new projects include the <Nullable>enable</Nullable> element in the project file.
  Update our project accordingly.

lvernaillen committed:
Move to implicit usings and file scoped namespaces

Reduce project complexity and make it more friendly for beginners
- Remove unneeded using statements because of enabled ImplicitUsings
- Change to file scoped namespace

lvernaillen commented:
I can move the api to minimal api if you want.
Do you want it in an separate PR or should I add it in this one?

Also, if you want to reduce project complexity and make it more friendly for beginners, do you want to keep a solution with the two projects _Starter.Api_ and _Starter.Core_?

I think it would be more accessible for beginners if there is no solution and only one project, similar to the dotnet example created from `dotnet new webapi -minimal -o SomeApi`.
I would remove the solution and the _Starter.Core_ project, leaving only the _Starter.Api_ project. The model classes from _Starter.Core_ could be moved to a _Model_ folder inside the _Starter.Api_ project.

neistow commented:
> Do you want it in an separate PR or should I add it in this one?

You can add changes to this PR.

> I think it would be more accessible for beginners if there is no solution and only one project, similar to the dotnet example created from `dotnet new webapi -minimal -o SomeApi`.

I see no harm in the solution. It can be beneficial for people, who really want to separate things or develop a starter template into something bigger.

> I would remove the solution and the _Starter.Core_ project, leaving only the _Starter.Api_ project. The model classes from _Starter.Core_ could be moved to a _Model_ folder inside the _Starter.Api_ project.

Agreed. You can remove `Starter.Core` and move models into `Starter.Api`, but leave the solution.

neistow commented on Starter.Api/Controllers/SnakeController.cs:
```diff
@@ -1,76 +1,72 @@
+    public IActionResult Move(GameStatusRequest gameStatusRequest)
+    {
+        var direction = new List<string> { "down", "left", "right", "up" };
+        var rng = new Random();
 
-            var response = new MoveResponse
-            {
-                Move = direction[rng.Next(direction.Count)],
-                Shout = "I am moving!"
-            };
-            return Ok(response);
-        }
+        var response = new MoveResponse
+        {
+            Move = direction[rng.Next(direction.Count)],
+            Shout = "I am moving!"
+        };
+        return Ok(response);
+    }
 
 
-        /// <summary>
-        /// Your Battlesnake will receive this request whenever a game it was playing has ended.
-        /// Use it to learn how your Battlesnake won or lost and deallocated any server-side resources.
-        /// Your response to this request will be ignored.
-        /// </summary>
-        [HttpPost("end")]
-        public IActionResult End(GameStatusRequest gameStatusRequest)
-        {
-            return Ok();
-        }
+    /// <summary>
+    /// Your Battlesnake will receive this request whenever a game it was playing has ended.
+    /// Use it to learn how your Battlesnake won or lost and deallocated any server-side resources.
+    /// Your response to this request will be ignored.
+    /// </summary>
+    [HttpPost("end")]
+    public IActionResult End(GameStatusRequest gameStatusRequest)
+    {
+        return Ok();
```
We can get rid of Controllers and use `MapPost` and equivalents in the `Program.cs`

neistow commented on Starter.Api/Program.cs:
```diff
@@ -1,17 +1,13 @@
-        {
-            CreateHostBuilder(args).Build().Run();
-        }
-
-        public static IHostBuilder CreateHostBuilder(string[] args) =>
-            Host.CreateDefaultBuilder(args)
-                .ConfigureWebHostDefaults(webBuilder => { webBuilder.UseStartup<Startup>(); });
+        CreateHostBuilder(args).Build().Run();
     }
+
+    public static IHostBuilder CreateHostBuilder(string[] args) =>
+        Host.CreateDefaultBuilder(args)
+            .ConfigureWebHostDefaults(webBuilder => { webBuilder.UseStartup<Startup>(); });
```
Same goes for old fashioned way of splitting app to `Program` and `Startup`

lvernaillen replied:
There is even no `Main(...)` method needed in the minimal api.
Just a sec. I'll push the update.

lvernaillen committed:
Move to minimal api

neistow commented on Starter.Api/Startup.cs:
```diff
@@ -1,41 +1,34 @@
-        }
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddControllers(o => { o.OutputFormatters.Clear(); })
+            .AddNewtonsoftJson(o => { o.UseCamelCasing(processDictionaryKeys: true); });
```
Also, I don't remember why I used `NewtonsoftJson` here 😅. `System.Text.Json` should [support](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions?view=net-6.0#:~:text=DictionaryKeyPolicy,as%20camel%2Dcasing.) setting casing both for properties and dictionary keys. But this should be 100% tested before implementing

lvernaillen replied:
There's no need to reference `NewtonsoftJson` or `System.Text.Json` anymore.
I tested the minimal api using the battlesnake CLI. See https://github.com/BattlesnakeOfficial/rules

lvernaillen committed:
Remove Starter.Core project

Make solution more simple for beginners.

lvernaillen committed:
Add missing initializer

neistow commented on Starter.Api/Starter.Api.csproj:
```diff
@@ -2,6 +2,8 @@
+        <ImplicitUsings>enable</ImplicitUsings>
     </PropertyGroup>
 
     <ItemGroup>
```
If we leave `NewtonsoftJson` package it should be also updated to the latest version

neistow commented on Starter.Api/Program.cs:
```diff
@@ -1,17 +1,65 @@
+app.MapPost("/move", (GameStatusRequest gameStatusRequest) =>
+{
+    var direction = new List<string> { "down", "left", "right", "up" };
+    var rng = new Random();
```
Better to use `Random.Shared`

neistow commented on Starter.Api/Program.cs:
```diff
@@ -1,17 +1,65 @@
+var builder = WebApplication.CreateBuilder(args);
+
+// Add services to the container.
+builder.Services.AddEndpointsApiExplorer();
```
Is `AddEndpointsApiExplorer()` necessary?

lvernaillen replied:
No it's not needed to run the minimal api.
It is needed if you'd want to add Swagger to it.
But since we don't use Swagger in this project, it might as well be removed.
You're right. I'll remove it.

lvernaillen committed:
Remove AddEndpointsApiExplorer

It is needed for Swagger for example, which we don't use in this project.

neistow reviewed (approved):
Look good to me

lvernaillen committed:
Use Random.Shared for thread safety
